File: app/src/main/java/com/effective/android/sample/MainActivity.java
package com.effective.android.sample;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;


import com.effective.android.anchors.LockableAnchor;
public class MainActivity extends AppCompatActivity {

    private LockableAnchor lockableAnchor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.test_user_anchor).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.d("MainActivity","Demo2 - start");
                lockableAnchor = new TaskTest().startForTestLockableAnchor();
                lockableAnchor.setLockListener(new LockableAnchor.LockListener() {
                    @Override
                    public void lockUp() {
                        //做一些自己的业务判断
                        //.....
                        new CusDialog.Builder(MainActivity.this)
                                .title("任务(" + lockableAnchor.getLockId() + ")已进入等待状态，请求响应")
                                .left("终止任务", new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        lockableAnchor.smash();
                                    }
                                })
                                .right("继续执行", new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        lockableAnchor.unlock();
                                    }
                                }).build().show();
                    }
                });
                Log.d("MainActivity","Demo2 - end");
            }
        });
    }


    static class CusDialog extends Dialog{

        private CusDialog(@NonNull Context context) {
            super(context);
        }

        public static class Builder{

            Context context;
            CharSequence title;
            CharSequence left;
            View.OnClickListener leftAction;
            CharSequence right;
            View.OnClickListener rightAction;
            boolean action = false;

            public Builder(Context context){
                this.context = context;
            }

            public Builder title(@NonNull CharSequence charSequence){
                if(!TextUtils.isEmpty(charSequence)){
                    this.title = charSequence;
                }
                return this;
            }

            public Builder left(@NonNull CharSequence charSequence){
                return left(charSequence,null);
            }

            public Builder left(@NonNull CharSequence charSequence, View.OnClickListener listener){
                if(!TextUtils.isEmpty(charSequence)){
                    this.left = charSequence;
                }
                leftAction = listener;
                return this;
            }

            public Builder right(@NonNull CharSequence charSequence){
                return right(charSequence,null);
            }

            public Builder right(@NonNull CharSequence charSequence, View.OnClickListener listener){
                if(!TextUtils.isEmpty(charSequence)){
                    this.right = charSequence;
                }
                rightAction = listener;
                return this;
            }

            public Dialog build(){
                final CusDialog cusDialog = new CusDialog(context);
                View view = LayoutInflater.from(context).inflate(R.layout.dialog_layout,null);
                TextView title = view.findViewById(R.id.title);
                title.setText(this.title);
                TextView left = view.findViewById(R.id.left);
                left.setText(this.left);
                left.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if(leftAction != null){
                            action = true;
                            leftAction.onClick(v);
                        }
                        cusDialog.dismiss();
                    }
                });
                TextView right = view.findViewById(R.id.right);
                right.setText(this.right);
                right.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if(rightAction != null){
                            action = true;
                            rightAction.onClick(v);
                        }
                        cusDialog.dismiss();
                    }
                });
                cusDialog.setContentView(view);
                Window window = cusDialog.getWindow();
                if (window != null) {
                    window.setGravity(Gravity.CENTER);
                    WindowManager.LayoutParams lp = window.getAttributes();
                    lp.width = WindowManager.LayoutParams.WRAP_CONTENT;
                    lp.height = WindowManager.LayoutParams.WRAP_CONTENT;
                    window.setAttributes(lp);
                }
                cusDialog.setCanceledOnTouchOutside(false);
                cusDialog.setCancelable(false);
                return cusDialog;
            }
        }
    }


}

File: app/src/main/java/com/effective/android/sample/view/MainActivity.kt
package com.effective.android.sample.view

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.effective.android.sample.R
import com.effective.android.sample.data.TaskTest
import com.effective.android.sample.util.ProcessUtils

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.d(TAG, "MainActivity#onCreate process Id is " + ProcessUtils.getProcessId())
        Log.d(TAG, "MainActivity#onCreate process Name is " + ProcessUtils.getProcessName())
        //SampleApplication#onCreate 已经有常规初始化的demo实例了
//留意跳转时Log.d 输出的新进程信息，同时application#onCreate会被再次调用
//一般针对多进程，在常规初始化流程中针对进程对应的特定场景构建特定的初始化链
//在异步进程被拉起的时候，执行特定进程的初始化链即可
//具体代码可参考 SampleApplication#initDependenciesCompatMutilProcess() 方法
        testPrivateProcess()
        testPublicProcess()
        //测试用户选择
        testUserChoose()
        //测试重启新链接
        testRestartNewDependenciesLink()
    }

    private fun testPrivateProcess() {
        findViewById<View>(R.id.test_private_process).setOnClickListener { startActivity(Intent(this@MainActivity, PrivateProcessActivity::class.java)) }
    }

    private fun testPublicProcess() {
        findViewById<View>(R.id.test_public_process).setOnClickListener { startActivity(Intent(this@MainActivity, PublicProcessActivity::class.java)) }
    }

    private fun testUserChoose() {
        findViewById<View>(R.id.test_user_anchor).setOnClickListener {
            Log.d("MainActivity", "Demo1 - testUserChoose")
            TaskTest().startForTestLockableAnchorByDsl {
                val lockableAnchor = it
                CusDialog.Builder(this@MainActivity)
                        .title("任务(" + lockableAnchor.lockId + ")已进入等待状态，请求响应")
                        .left("终止任务") { lockableAnchor.smash() }
                        .right("继续执行") { lockableAnchor.unlock() }.build().show()
            }
            Log.d("MainActivity", "Demo1 - testUserChoose")
        }
    }

    private fun testRestartNewDependenciesLink() {
        findViewById<View>(R.id.test_restart).setOnClickListener {
            Log.d("MainActivity", "Demo2 - testRestartNewDependenciesLink - startLinkOne")
            TaskTest().startForLinkOneByDsl(
                    Runnable {
                        Log.d("MainActivity", "Demo2 - testRestartNewDependenciesLink - endLinkOne")
                        Log.d("MainActivity", "Demo2 - testRestartNewDependenciesLink - startLinkTwo")
                        Handler(Looper.getMainLooper()).post {
                            TaskTest().startForLinkTwoByDsl()
                        }
                        Log.d("MainActivity", "Demo2 - testRestartNewDependenciesLink - endLinkTwo")
                    })
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}